### ISSUE TITLE
NNBD. No static warning to use the null check operator on strictly non-nullable expression

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#39867]
According to the NNBD [specification](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#static-errors) 

> It is a warning to use the null check operator (!) on an expression of type T if T is strictly non-nullable .

But there is no warning in dartanalyzer version 2.8.0-dev.0.0 for the following case

```dart
class A {}

main() {
  A? a = new A();
  if (a is Object) {
    a!;     //no static warning here
  }
}
```

**Comment by lrhn** [comment]
I don't think we promote to non-nullable on `is Object`.

We promote to non-nullable on `is Null`/`is! Null`/`== null`/`!= null`, but the `is Object` is considered as any other type check and only promote if `Object` is a subtype of the existing type of the variable being tested (which it isn't). @stereotype441 Am I right?

We *could* promote to non-nullable any time we do a successful `is` check against a definitely non-nullable type, even if we don't otherwise promote to the type.

**Comment by stereotype441** [comment]
> I don't think we promote to non-nullable on `is Object`.
> 
> We promote to non-nullable on `is Null`/`is! Null`/`== null`/`!= null`, but the `is Object` is considered as any other type check and only promote if `Object` is a subtype of the existing type of the variable being tested (which it isn't). @stereotype441 Am I right?

Yes, that matches my understanding.  Side note: promoting to non-nullable on `is Null`/`is! Null` doesn't currently work due to https://github.com/dart-lang/sdk/issues/40014.
